src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs

```diff
@@ -0,0 +1,163 @@
+        // excess semantic work.
+        var semanticModel = await document.GetRequiredNullableDisabledSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var seenDocumentIds = new ConcurrentSet<DocumentId>();
+        var seenTypeNames = new ConcurrentSet<string>();
+
+        // Bind as much as we can in parallel to get the results as quickly as possible.  We call into the
+        // ProducerConsumer overload that will batch up values into arrays, and call into the callback with them.  While
+        // that callback is executing, the ProducerConsumer continues to run, computing more results and getting them
+        // ready for when that call returns.  This approach ensures that we're not pausing while we're reporting the
+        // results to whatever client is calling into us.
+        await ProducerConsumer<DocumentId>.RunParallelAsync(
+            // Order the nodes by the distance from the requested position.
+            IteratePotentialTypeNodes(root).OrderBy(t => t.expression.SpanStart - position),
+            produceItems: static (tuple, callback, args, cancellationToken) =>
+            {
+                var (solution, semanticModel, seenDocumentIds, seenTypeNames, callbackAsync) = args;
+                ProduceItems(
+                    solution, semanticModel, tuple.expression, tuple.nameToken, callback, seenDocumentIds, seenTypeNames, cancellationToken);
+
+                return Task.CompletedTask;
+            },
+            consumeItems: static async (array, args, cancellationToken) =>
+            {
+                var (solution, semanticModel, seenDocumentIds, seenTypeNames, callbackAsync) = args;
+                await callbackAsync(array, cancellationToken).ConfigureAwait(false);
+            },
+            args: (solution, semanticModel, seenDocumentIds, seenTypeNames, callbackAsync),
+            cancellationToken).ConfigureAwait(false);
+
+        return;
+
+        static IEnumerable<(ExpressionSyntax expression, SyntaxToken nameToken)> IteratePotentialTypeNodes(SyntaxNode root)
+        {
+            using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var stack);
+            stack.Push(root);
+
+            while (stack.TryPop(out var current))
+            {
+                if (current is MemberAccessExpressionSyntax memberAccess)
+                {
+                    // Could be a static member access off of a type name.  Check the left side, and if it's just a
+                    // dotted name, return that.
+
+                    if (IsPossibleTypeName(memberAccess.Expression, out var nameToken))
+                    {
+                        // Something like `X.Y.Z` where `X.Y` is a type name.  Bind X.Y
+                        yield return (memberAccess.Expression, nameToken);
```

Comment:
For `X.Y.Z`, memberAccess.Expression is `X.Y`, but the nameToken returned would be `X`, right? Shouldn't it be `Y`?

Reply:
not exactly.  we have `X.Y.Z`.  But we pass `X.Y` Into IsPossibleTypeName, and if it succeeds, we return `Y`.